preliminary hacking up of httpgrpc middleware

francoposa opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

charleskorn commented on pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go:
```diff
@@ -0,0 +1,63 @@
+//
+// Root or not, the server Span will be embedded in the context.Context for the
+// application-specific gRPC handler(s) to access.
+func OpenTracingHTTPGRPCUnaryServerInterceptor(tracer opentracing.Tracer) grpc.UnaryServerInterceptor {
```
Do we need an interceptor for this? Or is this something that could be done directly in the handler for `httpgrpc` requests? (If so, this will likely require moving these changes to `weaveworks/common/httpgrpc`.)

francoposa replied:
I think setting tracing spans and metadata is much more appropriate for middleware/interceptors, rather than the handlers that are doing application logic.

But there are likely some enhancements that we could add to the library itself to make this integrate a little better

charleskorn replied:
Sorry, I should have been clearer - when I said "handler for `httpgrpc` requests", I meant the handler in the `httpgrpc` server (ie. [here](https://github.com/weaveworks/common/blob/dd9e68f319d5ddfb68002a238d481265af23fa3b/httpgrpc/server/server.go#L48)), not the application itself.

charleskorn commented on pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go:
```diff
@@ -0,0 +1,63 @@
+					span := opentracing.SpanFromContext(ctx)
+					span.SetTag("http.url", fullReqURL).SetTag("http.method", reqMethod)
+					span, ctx = opentracing.StartSpanFromContextWithTracer(
+						ctx, tracer, httpGRPCHandleMethod+parsedReqURL.Path,
```
Rather than creating a new span with the path in the name, could we change the name of the existing span (the one retrieved on line 49) to include the path? Looks like there's a `SetOperationName` method we could use to override it here.

charleskorn replied:
[nit] Minor nit on the span name: would be good to separate the gRPC method name and the path from the wrapped request. At the moment, we get something like `/httpgrpc.HTTP/Handle/v1/blah/foo`. Maybe something like `/httpgrpc.HTTP/Handle (/v1/blah/foo)` would be clearer? (Not too attached to the exact format)

charleskorn reviewed (commented):
Thanks for picking this up @francoposa!

Apart from the comments below, I have one other thought: should we do something similar for the client-side spans as well? Would be good to add similar information to those spans.

francoposa commented:
@charleskorn responding to all Q's here:

First of all, yes we should do it on the client side to; will start on that next.

Regarding span naming, I have been digging into the conventions here, mostly leaning on OpenTelemetry conventions because OpenTracing defined very little in the way of conventions and OpenTelemetry is the future state for tracing anyway.

This all comes with a little grain of salt regarding how we utilize these conventions because we have put ourselves in a unique and awkward spot with this library where not only are doing http-over-grpc, but we are also doing everything on the client and server through these general purpose `httpgrpc.HTTP/Handle` methods rather than one-method-per-RPC.

From OTEL RPC conventions: https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/rpc/

> Span name
> The span name MUST be the full RPC method name formatted as:
> 
> $package.$service/$method
> 
> (where $service MUST NOT contain dots and $method MUST NOT contain slashes)
> 
> If there is no package name or if it is unknown, the $package. part (including the period) is omitted.
> 
> Examples of span names:
> 
> `grpc.test.EchoService/Echo`
> `com.example.ExampleRmiService/exampleMethod`
> `MyCalcService.Calculator/Add` reported by the server and `MyServiceReference.ICalculator/Add` reported by the client for .NET WCF calls
> `MyServiceWithNoPackage/theMethod`

From OTEL HTTP conventions: https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/http/

> HTTP spans MUST follow the overall [guidelines for span names](https://opentelemetry.io/docs/specs/otel/trace/api/#span). HTTP server span names SHOULD be {http.request.method} {http.route} if there is a (low-cardinality) http.route available. HTTP server span names SHOULD be {http.method} if there is no (low-cardinality) http.route available. HTTP client spans have no http.route attribute since client-side instrumentation is not generally aware of the “route”, and therefore HTTP client spans SHOULD use {http.method}. Instrumentation MUST NOT default to using URI path as span name, but MAY provide hooks to allow custom logic to override the default span name.

And finally a note on changing the name of a span after it's already started:

> UpdateName
> 
> Updates the Span name. Upon this update, any sampling behavior based on Span name will depend on the implementation.
> 
> Note that [Samplers](https://opentelemetry.io/docs/specs/otel/trace/sdk/#sampler) can only consider information already present during span creation. Any changes done later, including updated span name, cannot change their decisions.
> 
> Alternatives for the name update may be late Span creation, when Span is started with the explicit timestamp from the past at the moment where the final Span name is known, or reporting a Span with the desired name as a child Span.

Given what I've read, I think:

1. Simply overriding the original `httpgrpc.HTTP/Handle` span name is ill-advised, this runs afoul of the RPC span naming conventions and has the issue of some undefined sampling behavior if you rename a span after it already started.

2. Only utilizing the span attributes to mark up the extra information, and not having a separate span name at all seems least-wrong given the conventions presented

3. If we absolutely want to have the HTTP route in the span name, then we combine the RPC and HTTP conventions. We take the required RPC method name from the convention, then add the HTTP route convention, basically taking these two bits:

RPC:
> Span name
> The span name MUST be the full RPC method name formatted as:
> 
> $package.$service/$method
> 
> (where $service MUST NOT contain dots and $method MUST NOT contain slashes)

and HTTP:

> HTTP server span names SHOULD be {http.request.method} {http.route} if there is a (low-cardinality) http.route available

and combining, we would get `httpgrpc.HTTP/Handle /v1/blah/foo`  - delimited just by a space in between, like the HTTP {http.request.method} {http.route} convention.

cyrille-leclerc commented:
Should the span name be `httpgrpc.HTTP/Handle /v1/blah/foo` or `httpgrpc.HTTP/Handle POST /v1/blah/foo`? Should it integrate `http.request.method`?

francoposa commented:
> Should the span name be [...] httpgrpc.HTTP/Handle POST /v1/blah/foo? Should it integrate http.request.method?

@cyrille-leclerc yes this is probably better and more consistent.

charleskorn commented:
I like @cyrille-leclerc's suggestion for the span name as well.

bboreham commented:
> Should the span name be `httpgrpc.HTTP/Handle /v1/blah/foo` or `httpgrpc.HTTP/Handle POST /v1/blah/foo`? Should it integrate `http.request.method`?

Suggest removing one 'http', and dropping 'Handle' which adds no meaning.
I.e. `httpgrpc /v1/blah/foo`.

I don't think POST vs GET is very interesting in the places we use this.

francoposa commented:
> Suggest removing one 'http', and dropping 'Handle' which adds no meaning.
I.e. httpgrpc /v1/blah/foo.

Changing the gRPC method name (which is `httpgrpc.HTTP/Handle`) would just be a full break from the semantic convention 

> I don't think POST vs GET is very interesting in the places we use this.

dropping the http method would also be a convention break

... though this is all with a grain of salt as we are just trying to do our best to map the conventions onto this nonstandard setup

francoposa commented:
So I have managed to also push the instrumentation down into the internal HTTP Handler that `httpgrpc` sends these gRPC-wrapped HTTP requests to after unwrapping them on the server side.

This actually gives us the opportunity to not break any conventions at all.

This approaches leaves the `httpgrpc.HTTP/Handle` span name in place, but still attaches all relevant http span tags to that span `http.request.method`, `http.route`, and `url.path`.

Then down in the internal http handler, we start a new span with a conventional HTTP middleware approach, following all the HTTP span conventions, as once we reach this layer we are truly only working with an HTTP request.
We get a standardized HTTP span name like `POST /otlp/v1/metrics`, and then again all standard span tags `http.request.method`, `http.route`, and `url.path`.

The downside here is that the `httpgrpc.HTTP/Handle` span _name_ is still so generic - but still, now it would have tags that can be filtered on.

Traces would look like this:
![Screenshot from 2023-08-01 11-04-42](https://github.com/grafana/mimir/assets/13006869/3921c75b-32f5-4842-b930-f9bd64c1d777)

charleskorn commented on pkg/mimir/mimir.go:
```diff
@@ -750,6 +752,10 @@ func New(cfg Config, reg prometheus.Registerer) (*Mimir, error) {
 	otel.SetTracerProvider(NewOpenTelemetryProviderBridge(opentracing.GlobalTracer()))
+	//middleware.InitGRPCMiddleware(&mimir.Cfg.Server)
+
+	mimir.Cfg.Server.Router = mux.NewRouter()
```
Is this router used for direct HTTP requests as well as handling `httpgrpc` requests? If so, do we need to do something in the middleware to only add the tags to the parent span and create the child span if we're handling a `httpgrpc` request?

francoposa replied:
Will look closer at this

charleskorn commented on pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go:
```diff
@@ -0,0 +1,116 @@
+			opentracing.Tag{Key: "http.route", Value: matchedRoute},
+			opentracing.Tag{Key: "url.path", Value: r.RequestURI},
+		}
+		childSpan, _ := opentracing.StartSpanFromContextWithTracer(ctx, tracer, childSpanName, startSpanOpts...)
+		defer childSpan.Finish()
+
+		next.ServeHTTP(w, r)
```
Do we need to do something here to pass the new context (with the new `childSpan` span) to `r`, so that spans created by `next` are nested under `childSpan`? Based on the screenshot you shared, it looks like child spans are nested under `parentSpan`, rather than `childSpan`.

francoposa replied:
I also noticed this and I'm going to poke at it a bit

charleskorn reviewed (commented):
HTTP middleware approach looks good to me.

francoposa commented:
@charleskorn I believe all points have been addressed; this latest commit fixes the parent-child relationship of the spans and ensures we only tag the parent span if it is `httpgrpc.HTTP/Handle`.

Open to paring down the docstrings if they are too much.

I'm inclined to punt the client-side version of this to another PR, if we still want it.

Traces look like this now:
![Screenshot from 2023-08-03 19-50-22](https://github.com/grafana/mimir/assets/13006869/1d7b2520-682b-400c-9d41-f6691875299e)

charleskorn commented on pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go:
```diff
@@ -0,0 +1,89 @@
+				parentSpan.SetTag("url.path", r.RequestURI)
+			}
+		}
+
+		childSpanName := httpServerSpanName(r.Method, matchedRoute)
+		startSpanOpts := []opentracing.StartSpanOption{
+			ext.SpanKindRPCServer,
+			opentracing.Tag{Key: string(ext.Component), Value: "net/http"},
+			opentracing.Tag{Key: "http.request.method", Value: r.Method},
+			opentracing.Tag{Key: "http.route", Value: matchedRoute},
+			opentracing.Tag{Key: "url.path", Value: r.RequestURI},
+		}
+		childSpan, _ := opentracing.StartSpanFromContextWithTracer(ctx, tracer, childSpanName, startSpanOpts...)
+		defer childSpan.Finish()
```
What happens if a HTTP request arrives directly (ie. not through httpgrpc)? 

Do we end up with a duplicate span? (I'd expect standard HTTP requests to already have a span created for them.)

francoposa replied:
So it does create two HTTP spans, though I think this may actually be useful because we will get the same span child span regardless of whether the parent is an HTTP or gRPC request.

The parent HTTP span is not identical to the child HTTP span we're creating, because it's using default opentracing middleware, which adds some info like the status code and user agent (and opentracing has a different span naming convention).

charleskorn replied:
I think it would be confusing to see two spans for the non-httpgrpc case, so only emitting the span when it's a httpgrpc request makes sense to me.

>The parent HTTP span is not identical to the child HTTP span we're creating, because it's using default opentracing middleware, which adds some info like the status code and user agent (and opentracing has a different span naming convention).

Should we also attach similar info? I imagine the status code in particular would be very useful.

charleskorn commented on pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go:
Would be good to add a test for the functionality here.

charleskorn reviewed (commented):
Generally looks good to me. I'd suggest moving this to `grafana/dskit` so that all applications that use `httpgrpc` can benefit from this.

francoposa commented:
> I'd suggest moving this to grafana/dskit so that all applications that use httpgrpc can benefit from this.

I think it may be better to wait until Mimir is actually migrated from weaveworks/common to the grafana/dskit libs, to avoid introducing changes in what should be a drop-in replacement change.

charleskorn commented:
> > I'd suggest moving this to grafana/dskit so that all applications that use httpgrpc can benefit from this.
> 
> I think it may be better to wait until Mimir is actually migrated from weaveworks/common to the grafana/dskit libs, to avoid introducing changes in what should be a drop-in replacement change.
> 
> We could push this into Mimir now, where it will work regardless of whether the server is created from weaveworks/common of grafana/dskit, then migrate it into dskit later OR we can just wait until dskit is in Mimir, then create this directly in dskit.

Mimir has already migrated 🙂 https://github.com/grafana/mimir/pull/5672

francoposa commented:
@charleskorn with that feedback, I will plan to move this over to a dskit PR, with the following changes:

1. Emit HTTP span only when parent is `httpgrpc.HTTP/Handle`, not already an HTTP span
2. Emulate the existing HTTP spans that are created when it _does_ come through directly as an HTTP request
3. Write a test - not familiar with how to unit test tracing span emission, but will start by checking what exists for the opentracing libraries we already use in Mimir

charleskorn commented:
Sounds great, thanks @francoposa!

francoposa commented:
moved to https://github.com/grafana/dskit/pull/353 if anyone wants to take a look at what we ended up with; moved back to OpenTracing style rather than OTEL for now to match what is currently in the other middlewares
